src/audio/kpb.c

```diff
@@ -1050,6 +1051,8 @@ static enum task_state kpb_draining_task(void *arg)
 
 	trace_kpb("kpb_draining_task(), start.");
 
+	pm_runtime_disable(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID);
+
```

Comment:
The commit message and expected transitions are confusing to me.
If you want to prevent power gating while you are draining, then this code looks ok, but the state machine would be very odd. How could this even happen?
If you want to wake and exit power-gated mode, then how would this result in any action: there's nothing running.
If you want to make sure the drain happens at full speed, then how could power gating get in the way?

Reply:
> The commit message and expected transitions are confusing to me.
> If you want to prevent power gating while you are draining, then this code looks ok, but the state machine would be very odd. How could this even happen?

It is very simple, at the boot time, pm-runtime is disabled by the platform initialization routine (sref=1) and the driver is aligned. Once the driver decides, a longer latency of int response is acceptable (and no large IPCs which require mailbox access are expected), may enable the pm-runtime by sending PM_GATE(PPG=0) which lowers the cnt to 0 and platform begins to power gate the dsp in waiti. Any time the driver wants to bring back the dsp to full speed and enable all IPCs capabilities, sends the PM_GATE (PPG=1).

> If you want to wake and exit power-gated mode, then how would this result in any action: there's nothing running.
> If you want to make sure the drain happens at full speed, then how could power gating get in the way?

It just prevents to power gate in waiti (which may happen if progress on the capture buffer is int-driven). power-gated dsp has a longer wake up time when an interrupt comes, we want to avoid that.  sref goes from 0 to 1 between PM_GATE (PPG=0) and PM_GATE(PPG=1).

Reply:
> > The commit message and expected transitions are confusing to me.
> > If you want to prevent power gating while you are draining, then this code looks ok, but the state machine would be very odd. How could this even happen?
> 
> It is very simple, at the boot time, pm-runtime is disabled by the platform initialization routine (sref=1) and the driver is aligned. Once the driver decides, a longer latency of int response is acceptable (and no large IPCs which require mailbox access are expected), may enable the pm-runtime by sending PM_GATE(PPG=0) which lowers the cnt to 0 and platform begins to power gate the dsp in waiti. Any time the driver wants to bring back the dsp to full speed and enable all IPCs capabilities, sends the PM_GATE (PPG=1).

right, but as in my other comments pm_runtime_enable/disable isn't supposed to be refcounted.

> 
> > If you want to wake and exit power-gated mode, then how would this result in any action: there's nothing running.
> > If you want to make sure the drain happens at full speed, then how could power gating get in the way?
> 
> It just prevents to power gate in waiti (which may happen if progress on the capture buffer is int-driven). power-gated dsp has a longer wake up time when an interrupt comes, we want to avoid that. sref goes from 0 to 1 between PM_GATE (PPG=0) and PM_GATE(PPG=1).

I get what you are trying to do but that's a get/put operation to prevent pm_runtime from disabling a resource you need. It's different from enable/disable.